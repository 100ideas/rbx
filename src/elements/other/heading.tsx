import { cx } from "emotion";
import React from "react";

import { forwardRefAs } from "@/generic";
import { ModifierProps, transformModifiers } from "@/modifiers";

export type HeadingProps = ModifierProps;

export const Heading = forwardRefAs<HeadingProps, "p">((props, ref) => {
  const { as, ...rest } = transformModifiers(props);
  rest.className = cx("heading", rest.className);
  return React.createElement(as!, { ref, ...rest });
}, "p");
