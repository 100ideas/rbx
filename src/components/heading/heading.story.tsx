import React from "react";

import { storiesOf } from "@storybook/react";

import { Box } from "components/box";
import { Heading } from "components/heading";

storiesOf("Heading", module).add("Default", () => (
  <div>
    <Box>
      <Heading>Title</Heading>
      <Heading subtitle size={6 as 6}>
        Subtitle
      </Heading>
    </Box>
    <Box>
      <Heading size={1 as 1}>Title</Heading>
      <Heading subtitle size={3 as 3}>
        Subtitle
      </Heading>
    </Box>
    <Box>
      <Heading size={2 as 2}>Title</Heading>
      <Heading subtitle size={4} renderAs="h2">
        Subtitle
      </Heading>
    </Box>
    <Box>
      <Heading size={3 as 3}>Title</Heading>
      <Heading subtitle size={5} renderAs="h2">
        Subtitle
      </Heading>
    </Box>
    <Box>
      <Heading size={4 as 4}>Title</Heading>
      <Heading subtitle size={6} renderAs="h2">
        Subtitle
      </Heading>
    </Box>
  </div>
));
