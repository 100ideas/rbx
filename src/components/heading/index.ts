import "./heading.sass";

export { Heading } from "./heading";
