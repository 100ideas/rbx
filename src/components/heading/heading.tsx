import { cx } from "emotion";
import React from "react";

import { asExoticComponent } from "@/components/exotic";
import { ModifierProps, transformModifiers } from "@/modifiers";

export type HeadingModifierProps = Partial<{
  heading: boolean;
  size: 1 | 2 | 3 | 4 | 5 | 6;
  spaced: boolean;
  subtitle: boolean;
  weight: "light" | "normal" | "semibold" | "bold";
}>;

export type HeadingProps = ModifierProps & HeadingModifierProps;

export const Heading = asExoticComponent<HeadingProps, "h1">((props, ref) => {
  const {
    as,
    heading,
    size,
    spaced,
    subtitle,
    weight,
    ...rest
  } = transformModifiers(props);
  rest.className = cx(rest.className, {
    [`has-text-weight-${weight}`]: weight,
    heading,
    [`is-${size}`]: !!size,
    "is-spaced": spaced && !subtitle,
    subtitle,
    title: !subtitle && !heading,
  });
  return React.createElement(as!, { ref, ...rest });
}, "h1");

Heading.defaultProps = Object.assign(
  {
    heading: false,
    spaced: false,
    subtitle: false,
  },
  Heading.defaultProps,
);
