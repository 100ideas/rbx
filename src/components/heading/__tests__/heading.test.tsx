import React from "react";
import renderer from "react-test-renderer";

import { Heading } from "../heading";

describe("Heading component", () => {
  it("should exist", () => {
    expect(Heading).toMatchSnapshot();
  });

  it("should have title classname", () => {
    const component = renderer.create(
      <Heading>
        Test <a>Give me</a>
      </Heading>,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("should concat classname in props with Bulma classname", () => {
    const component = renderer.create(
      <Heading className="other-class this-is-a-test">
        <p>Default</p>
      </Heading>,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("should use inline styles", () => {
    const component = renderer.create(
      <Heading style={{ height: 250 }}>
        <p>Default</p>
      </Heading>,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("should be a subtitle with size rendered as P", () => {
    const component = renderer.create(
      <Heading<"p"> size={3} subtitle as="p">
        Subtitle
      </Heading>,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("should not be spaced because is subtitle", () => {
    const component = renderer.create(
      <Heading<"p"> spaced subtitle as="p">
        Subtitle
      </Heading>,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("should be spaced", () => {
    const component = renderer.create(
      <Heading<"p"> spaced as="p">
        Subtitle
      </Heading>,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});
