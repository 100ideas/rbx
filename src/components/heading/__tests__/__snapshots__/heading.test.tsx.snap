// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Heading component should be a subtitle with size rendered as P 1`] = `
<p
  className="is-3 subtitle"
>
  Subtitle
</p>
`;

exports[`Heading component should be spaced 1`] = `
<p
  className="is-spaced title"
>
  Subtitle
</p>
`;

exports[`Heading component should concat classname in props with Bulma classname 1`] = `
<h1
  className="other-class this-is-a-test title"
>
  <p>
    Default
  </p>
</h1>
`;

exports[`Heading component should exist 1`] = `
Object {
  "$$typeof": Symbol(react.forward_ref),
  "defaultProps": Object {
    "as": "h1",
    "heading": false,
    "spaced": false,
    "subtitle": false,
  },
  "render": [Function],
}
`;

exports[`Heading component should have title classname 1`] = `
<h1
  className="title"
>
  Test 
  <a>
    Give me
  </a>
</h1>
`;

exports[`Heading component should not be spaced because is subtitle 1`] = `
<p
  className="subtitle"
>
  Subtitle
</p>
`;

exports[`Heading component should use inline styles 1`] = `
<h1
  className="title"
  style={
    Object {
      "height": 250,
    }
  }
>
  <p>
    Default
  </p>
</h1>
`;
